File: sidecars/sih_workflow/draft.py
"""Typed ApprovalNoteDraft and related data structures.

These are the *only* structures the model is allowed to fill in. Trusted local
code (this package's renderer, validator, and verifier) reads them and either
turns them into a Word document or refuses to. The model never writes DOCX XML
directly.

Why typed:

* **Required fields are present or the draft is not renderable.** A missing
  equipment_id, an empty findings list, or a blank proposed action is a draft
  error caught before any file is written.
* **Every figure is bound to a calculation record.** A bare number in the
  proposed action cannot slip through; the verifier cross-checks it against the
  calculation log.
* **Every claim is bound to evidence.** An evidence_id that does not resolve
  to a passage in the authorized collection blocks readiness.

The choice of dataclasses with primitive types rather than free-form dicts is
the whole point of the typed intermediate: the model cannot smuggle a new
field through to the document by calling it something creative.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field, fields
from enum import Enum
from typing import Any, Dict, List, Optional


class DraftStatus(str, Enum):
    """Lifecycle of a draft.

    DRAFT is the only state in which a model may hand a draft to the renderer.
    Anything past DRAFT is set by trusted local code as a record of what
    happened next.
    """

    DRAFT = "DRAFT"
    PENDING_APPROVAL = "PENDING_APPROVAL"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    RENDERED = "RENDERED"


class Severity(str, Enum):
    """Finding severity, ordered low to critical."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

    @classmethod
    def order(cls) -> List["Severity"]:
        return [cls.LOW, cls.MEDIUM, cls.HIGH, cls.CRITICAL]

    def rank(self) -> int:
        return self.order().index(self)

    def label(self) -> str:
        return self.value.capitalize()


class Classification(str, Enum):
    """Material sensitivity.

    Ordered so a higher value (more sensitive) can be required to be matched or
    exceeded at every hop. The verifier uses this ordering to refuse a draft
    that would downgrade the output classification below the input.
    """

    PUBLIC = "public"
    INTERNAL = "internal"
    PROCESS_DIAGRAM = "processDiagram"
    CONFIDENTIAL = "confidential"
    RESTRICTED = "restricted"

    @classmethod
    def order(cls) -> List["Classification"]:
        return [
            cls.PUBLIC,
            cls.INTERNAL,
            cls.PROCESS_DIAGRAM,
            cls.CONFIDENTIAL,
            cls.RESTRICTED,
        ]

    def rank(self) -> int:
        return self.order().index(self)

    def at_least(self, minimum: "Classification") -> bool:
        return self.rank() >= minimum.rank()


@dataclass
class Finding:
    """A single finding the inspection produced."""

    id: str
    description: str
    severity: Severity
    location: Optional[str] = None
    source_page: Optional[int] = None
    evidence_ids: List[str] = field(default_factory=list)


@dataclass
class CalculationRef:
    """Pointer from a figure in the draft to the calculation record that
    produced it.

    The model never invents a number: every numerical figure the draft
    mentions is bound, by ID, to a record from the deterministic calculation
    engine. The verifier cross-checks this binding.
    """

    calculation_id: str
    expression: str
    result: str  # result with units, exactly as the engine returned it
    finding_id: Optional[str] = None


@dataclass
class EvidenceRef:
    """Pointer to a passage in the authorized collection.

    An evidence_id is the model-side label (``E1``, ``E2``...). The verifier
    resolves it against the loaded passage index, and refuses if the ID is not
    in the index.
    """

    evidence_id: str
    document_sha256: str
    page: int
    passage_text: str


@dataclass
class UncertaintyNote:
    """Something the inspection could not establish with certainty.

    ``blocks_approval`` distinguishes a missing value the verifier can
    work around from one that means the note cannot be signed until the gap
    is filled.
    """

    what: str
    reason: str
    finding_id: Optional[str] = None
    blocks_approval: bool = False


@dataclass
class ApprovalNoteDraft:
    """The typed intermediate between the model and the renderer.

    The model fills this in. The renderer reads it. No other channel is
    used for the content of the note.

    The required fields are exactly the ones the problem statement lists. A
    field is "filled" if it is present and non-empty; an empty string for a
    string field or an empty list for a list field counts as missing for the
    purposes of validation.
    """

    # --- Required fields, in the order the problem statement names them ---
    equipment_id: str
    inspection_date: str  # ISO 8601, e.g. 2026-08-26
    findings: List[Finding] = field(default_factory=list)
    severity: Severity = Severity.LOW
    evidence_ids: List[str] = field(default_factory=list)
    proposed_action: str = ""
    calculation_ids: List[str] = field(default_factory=list)
    uncertainty_notes: List[UncertaintyNote] = field(default_factory=list)
    classification: Classification = Classification.INTERNAL
    status: DraftStatus = DraftStatus.DRAFT

    # --- Provenance, set by the orchestrator, not the model ---
    model_id: Optional[str] = None
    skill_id: Optional[str] = None

    # ---------------------------------------------------------------------

    def to_dict(self) -> Dict[str, Any]:
        """Plain dict for JSON serialisation, with enums as their values."""
        out: Dict[str, Any] = {}
        for f in fields(self):
            value = getattr(self, f.name)
            if isinstance(value, Enum):
                out[f.name] = value.value
            elif isinstance(value, list) and value and isinstance(value[0], Finding):
                out[f.name] = [asdict(x) for x in value]
            elif isinstance(value, list) and value and isinstance(
                value[0], UncertaintyNote
            ):
                out[f.name] = [asdict(x) for x in value]
            else:
                out[f.name] = value
        return out

    def to_json(self) -> str:
        """Stable JSON for hashing. The renderer does not use this directly;
        the verifier hashes it to bind the approval to a specific draft."""
        return json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))

    def highest_severity(self) -> Severity:
        """Maximum severity across findings, or LOW when there are none."""
        if not self.findings:
            return Severity.LOW
        return max((f.severity for f in self.findings), key=lambda s: s.rank())

    def has_blocking_uncertainty(self) -> bool:
        return any(u.blocks_approval for u in self.uncertainty_notes)

    @classmethod
    def from_dict(cls, payload: Dict[str, Any]) -> "ApprovalNoteDraft":
        """Build a draft from a plain dict (e.g. JSON from a model).

        Unknown keys are dropped, on purpose: a model that invents a field
        cannot smuggle it into the renderer.
        """
        allowed = {f.name for f in fields(cls)}
        clean = {k: v for k, v in payload.items() if k in allowed}

        if "findings" in clean:
            clean["findings"] = [
                Finding(**_coerce_finding(f)) for f in clean["findings"]
            ]
        if "uncertainty_notes" in clean:
            clean["uncertainty_notes"] = [
                UncertaintyNote(**u) for u in clean["uncertainty_notes"]
            ]
        if "severity" in clean and isinstance(clean["severity"], str):
            clean["severity"] = Severity(clean["severity"])
        if "classification" in clean and isinstance(clean["classification"], str):
            clean["classification"] = Classification(clean["classification"])
        if "status" in clean and isinstance(clean["status"], str):
            clean["status"] = DraftStatus(clean["status"])
        return cls(**clean)


def _coerce_finding(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Coerce one finding dict's enum string into the enum value."""
    out = dict(payload)
    if isinstance(out.get("severity"), str):
        out["severity"] = Severity(out["severity"])
    return out

File: src-tauri/src/sih_workflow/draft.rs
//! Typed ApprovalNoteDraft structure.
//!
//! The model fills this typed draft. Trusted local code validates it and
//! renders the Word document. The model never writes DOCX XML directly.

use serde::{Deserialize, Serialize};

use crate::policy::Classification;

/// Status of the draft.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DraftStatus {
    Draft,
    PendingApproval,
    Approved,
    Rejected,
    Rendered,
}

/// Severity level of a finding.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Low,
    Medium,
    High,
    Critical,
}

impl Severity {
    pub fn label(self) -> &'static str {
        match self {
            Severity::Low => "Low",
            Severity::Medium => "Medium",
            Severity::High => "High",
            Severity::Critical => "Critical",
        }
    }
}

/// A single finding from the inspection report.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Finding {
    /// Unique ID for this finding within the draft.
    pub id: String,
    /// Description of what was found.
    pub description: String,
    /// Severity of the finding.
    pub severity: Severity,
    /// Optional location on the equipment.
    pub location: Option<String>,
    /// Page in the source report where this was found.
    pub source_page: Option<u32>,
    /// Evidence IDs that support this finding.
    pub evidence_ids: Vec<String>,
}

/// Reference to a calculation record.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CalculationRef {
    /// The calculation ID from the calculation engine.
    pub calculation_id: String,
    /// The expression that was evaluated.
    pub expression: String,
    /// The result with units.
    pub result: String,
    /// Which finding this calculation supports.
    pub finding_id: Option<String>,
}

/// Reference to an authorized evidence passage.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceRef {
    /// The evidence ID (matches a passage in the knowledge base).
    pub evidence_id: String,
    /// Document SHA-256.
    pub document_sha256: String,
    /// Page number in the source document.
    pub page: u32,
    /// The text of the passage.
    pub passage_text: String,
}

/// An uncertainty note from the extraction process.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UncertaintyNote {
    /// What is uncertain.
    pub what: String,
    /// Why it is uncertain.
    pub reason: String,
    /// Which finding this uncertainty applies to.
    pub finding_id: Option<String>,
    /// Whether this should block approval.
    pub blocks_approval: bool,
}

/// The typed ApprovalNoteDraft that the model fills in.
///
/// All fields are required to be present (non-null) for the draft to be
/// valid for rendering, even if empty. The `status` field tracks the
/// lifecycle of the draft.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalNoteDraft {
    /// Equipment ID being inspected.
    pub equipment_id: String,
    /// Date of inspection (ISO 8601).
    pub inspection_date: String,
    /// Findings from the inspection.
    pub findings: Vec<Finding>,
    /// Overall severity (highest severity among findings).
    pub severity: Severity,
    /// Evidence IDs from authorized passages.
    pub evidence_ids: Vec<String>,
    /// Proposed action based on the findings.
    pub proposed_action: String,
    /// References to calculation records.
    pub calculation_ids: Vec<String>,
    /// Uncertainty notes from the extraction.
    pub uncertainty_notes: Vec<UncertaintyNote>,
    /// Classification of the material.
    pub classification: Classification,
    /// Status of the draft.
    pub status: DraftStatus,
    /// Optional: model ID that produced this draft.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
    /// Optional: skill ID that guided this draft.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub skill_id: Option<String>,
}

impl ApprovalNoteDraft {
    /// Creates a new draft with the DRAFT status.
    pub fn new(
        equipment_id: String,
        inspection_date: String,
        classification: Classification,
    ) -> Self {
        Self {
            equipment_id,
            inspection_date,
            findings: Vec::new(),
            severity: Severity::Low,
            evidence_ids: Vec::new(),
            proposed_action: String::new(),
            calculation_ids: Vec::new(),
            uncertainty_notes: Vec::new(),
            classification,
            status: DraftStatus::Draft,
            model_id: None,
            skill_id: None,
        }
    }

    /// Returns the highest severity among all findings.
    pub fn highest_severity(&self) -> Severity {
        self.findings
            .iter()
            .map(|f| f.severity)
            .max()
            .unwrap_or(Severity::Low)
    }

    /// Returns true if any uncertainty note blocks approval.
    pub fn has_blocking_uncertainty(&self) -> bool {
        self.uncertainty_notes.iter().any(|u| u.blocks_approval)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new_draft_has_draft_status() {
        let draft = ApprovalNoteDraft::new(
            "EQ-001".to_string(),
            "2026-01-15".to_string(),
            Classification::Internal,
        );
        assert_eq!(draft.status, DraftStatus::Draft);
        assert_eq!(draft.equipment_id, "EQ-001");
    }

    #[test]
    fn highest_severity_returns_max() {
        let mut draft = ApprovalNoteDraft::new(
            "EQ-001".to_string(),
            "2026-01-15".to_string(),
            Classification::Internal,
        );
        draft.findings = vec![
            Finding {
                id: "F1".to_string(),
                description: "Minor corrosion".to_string(),
                severity: Severity::Low,
                location: None,
                source_page: Some(1),
                evidence_ids: vec![],
            },
            Finding {
                id: "F2".to_string(),
                description: "Crack detected".to_string(),
                severity: Severity::High,
                location: None,
                source_page: Some(2),
                evidence_ids: vec![],
            },
        ];
        assert_eq!(draft.highest_severity(), Severity::High);
    }

    #[test]
    fn has_blocking_uncertainty_checks_notes() {
        let mut draft = ApprovalNoteDraft::new(
            "EQ-001".to_string(),
            "2026-01-15".to_string(),
            Classification::Internal,
        );
        draft.uncertainty_notes = vec![
            UncertaintyNote {
                what: "Page 3 unclear".to_string(),
                reason: "Low OCR confidence".to_string(),
                finding_id: None,
                blocks_approval: false,
            },
        ];
        assert!(!draft.has_blocking_uncertainty());

        draft.uncertainty_notes.push(UncertaintyNote {
            what: "Critical value unreadable".to_string(),
            reason: "Scan quality too low".to_string(),
            finding_id: Some("F1".to_string()),
            blocks_approval: true,
        });
        assert!(draft.has_blocking_uncertainty());
    }
}
